lib/average_end_of_day_balance/main.py

```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env python3
+
+import os
+from datetime import datetime, timedelta
+from typing import Optional
+
+import pandas as pd
+from pngme.api import Client
+
+
+def get_average_end_of_day_balance(
+    client: Client, user_uuid: str, utc_starttime: datetime, utc_endtime: datetime
+) -> Optional[float]:
+    """Calculates the average end-of-day total balance for a user across all
+       of their accounts for a given time window.
+
+       Typical date ranges are last 30 days, 31-60 days and 61-90 days.
+       If no balance data was found, return None.
+
+    Args:
+        client: Pngme API client
+        user_uuid: the Pngme user_uuid for the mobile phone user
+        utc_starttime: the datetime for the left-hand-side of the time-window
+        utc_endtime: the datetime for the right-hand-side of the time-window
+
+    Returns:
+        the average end-of-day total balance for a given time window
+    """
+    institutions = client.institutions.get(user_uuid)
+    balance_df_list = []
+
+    """
+    Default Pngme balance API without time range input returns the latest balance data for an institution
+    Construct a time range in the API request to fetch all historical balance data
+    """
+    today = datetime.today()
+    all_pages_utc_endtime = datetime(today.year, today.month, today.day)
+    all_pages_starttime = all_pages_utc_endtime - timedelta(days=1e5)
+
+    for institution in institutions:
+        institution_id = institution.institution_id
+        institution_balance_records = client.balances.get(
+            user_uuid=user_uuid,
+            institution_id=institution_id,
+            utc_starttime=all_pages_starttime,
+            utc_endtime=all_pages_utc_endtime,
+        )
+        depository_records = []
+        for balance_record in institution_balance_records:
+            if balance_record.account_type == "depository":
+                balance_record_dict = balance_record.dict()
+                depository_records.append(
+                    {
+                        key: balance_record_dict[key]
+                        for key in ["account_number", "ts", "balance"]
+                    }
+                )
+        institution_balances_df = pd.DataFrame(depository_records)
+        institution_balances_df = institution_balances_df.assign(
+            institution_name=institution_id
+        )
+        balance_df_list.append(institution_balances_df)
+
+    # 1. Combine all institution balances into single dataframe
+    balances_df = pd.concat(balance_df_list)
```

Comment:
you can `del balance_df_list` after this operation if you want; pd dataframes are heavy, so worth dropping temporary dfs whenever possible

Reply:
I felt this should be properly handled by the python garbage collector, so will add this in now.